// payload/dist/admin/components/forms/field-types/RichText/elements/upload/Element/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const slate_1 = require("slate");
const slate_react_1 = require("slate-react");
const react_i18next_1 = require("react-i18next");
const Config_1 = require("../../../../../../utilities/Config");
const usePayloadAPI_1 = __importDefault(require("../../../../../../../hooks/usePayloadAPI"));
const File_1 = __importDefault(require("../../../../../../graphics/File"));
const useThumbnail_1 = __importDefault(require("../../../../../../../hooks/useThumbnail"));
const Button_1 = __importDefault(require("../../../../../../elements/Button"));
const getTranslation_1 = require("../../../../../../../../utilities/getTranslation");
const DocumentDrawer_1 = require("../../../../../../elements/DocumentDrawer");
const ListDrawer_1 = require("../../../../../../elements/ListDrawer");
require("./index.scss");
const baseClass = 'rich-text-upload';
const initialParams = {
    depth: 0,
};
const Element = ({ attributes, children, element }) => {
    var _a, _b, _c;
    const { relationTo, value, fieldProps, } = element;
    const { collections, serverURL, routes: { api } } = (0, Config_1.useConfig)();
    const { t, i18n } = (0, react_i18next_1.useTranslation)('fields');
    const [cacheBust, dispatchCacheBust] = (0, react_1.useReducer)((state) => state + 1, 0);
    const [relatedCollection, setRelatedCollection] = (0, react_1.useState)(() => collections.find((coll) => coll.slug === relationTo));
    const [ListDrawer, ListDrawerToggler, { closeDrawer: closeListDrawer, },] = (0, ListDrawer_1.useListDrawer)({
        uploads: true,
        selectedCollection: relatedCollection.slug,
    });
    const [DocumentDrawer, DocumentDrawerToggler, { closeDrawer, },] = (0, DocumentDrawer_1.useDocumentDrawer)({
        collectionSlug: relatedCollection.slug,
        id: value === null || value === void 0 ? void 0 : value.id,
    });
    const editor = (0, slate_react_1.useSlateStatic)();
    const selected = (0, slate_react_1.useSelected)();
    const focused = (0, slate_react_1.useFocused)();
    // Get the referenced document
    const [{ data }, { setParams }] = (0, usePayloadAPI_1.default)(`${serverURL}${api}/${relatedCollection.slug}/${value === null || value === void 0 ? void 0 : value.id}`, { initialParams });
    const thumbnailSRC = (0, useThumbnail_1.default)(relatedCollection, data);
    const removeUpload = (0, react_1.useCallback)(() => {
        const elementPath = slate_react_1.ReactEditor.findPath(editor, element);
        slate_1.Transforms.removeNodes(editor, { at: elementPath });
    }, [editor, element]);
    const updateUpload = (0, react_1.useCallback)((json) => {
        const { doc } = json;
        const newNode = {
            fields: doc,
        };
        const elementPath = slate_react_1.ReactEditor.findPath(editor, element);
        slate_1.Transforms.setNodes(editor, newNode, { at: elementPath });
        // setRelatedCollection(collections.find((coll) => coll.slug === collectionConfig.slug));
        setParams({
            ...initialParams,
            cacheBust, // do this to get the usePayloadAPI to re-fetch the data even though the URL string hasn't changed
        });
        dispatchCacheBust();
        closeDrawer();
    }, [editor, element, setParams, cacheBust, closeDrawer]);
    const swapUpload = react_1.default.useCallback(({ docID, collectionConfig }) => {
        const newNode = {
            type: 'upload',
            value: { id: docID },
            relationTo: collectionConfig.slug,
            children: [
                { text: ' ' },
            ],
        };
        const elementPath = slate_react_1.ReactEditor.findPath(editor, element);
        setRelatedCollection(collections.find((coll) => coll.slug === collectionConfig.slug));
        slate_1.Transforms.setNodes(editor, newNode, { at: elementPath });
        dispatchCacheBust();
        closeListDrawer();
    }, [closeListDrawer, editor, element, collections]);
    return (react_1.default.createElement("div", { className: [
            baseClass,
            (selected && focused) && `${baseClass}--selected`,
        ].filter(Boolean).join(' '), contentEditable: false, ...attributes },
        react_1.default.createElement("div", { className: `${baseClass}__card` },
            react_1.default.createElement("div", { className: `${baseClass}__topRow` },
                react_1.default.createElement("div", { className: `${baseClass}__thumbnail` }, thumbnailSRC ? (react_1.default.createElement("img", { src: thumbnailSRC, alt: data === null || data === void 0 ? void 0 : data.filename })) : (react_1.default.createElement(File_1.default, null))),
                react_1.default.createElement("div", { className: `${baseClass}__topRowRightPanel` },
                    react_1.default.createElement("div", { className: `${baseClass}__collectionLabel` }, (0, getTranslation_1.getTranslation)(relatedCollection.labels.singular, i18n)),
                    react_1.default.createElement("div", { className: `${baseClass}__actions` },
                        (value === null || value === void 0 ? void 0 : value.id) && (react_1.default.createElement(DocumentDrawerToggler, { className: `${baseClass}__toggler` },
                            react_1.default.createElement(Button_1.default, { icon: "edit", round: true, buttonStyle: "icon-label", el: "div", className: `${baseClass}__actionButton`, onClick: (e) => {
                                    e.preventDefault();
                                }, tooltip: t('general:editLabel', { label: relatedCollection.labels.singular }), disabled: (_a = fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.admin) === null || _a === void 0 ? void 0 : _a.readOnly }))),
                        react_1.default.createElement(ListDrawerToggler, null,
                            react_1.default.createElement(Button_1.default, { icon: "swap", round: true, buttonStyle: "icon-label", className: `${baseClass}__actionButton`, onClick: () => {
                                    // do nothing
                                }, el: "div", tooltip: t('swapUpload'), disabled: (_b = fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.admin) === null || _b === void 0 ? void 0 : _b.readOnly })),
                        react_1.default.createElement(Button_1.default, { icon: "x", round: true, buttonStyle: "icon-label", className: `${baseClass}__actionButton`, onClick: (e) => {
                                e.preventDefault();
                                removeUpload();
                            }, tooltip: t('removeUpload'), disabled: (_c = fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.admin) === null || _c === void 0 ? void 0 : _c.readOnly })))),
            react_1.default.createElement("div", { className: `${baseClass}__bottomRow` },
                react_1.default.createElement("strong", null, data === null || data === void 0 ? void 0 : data.filename))),
        children,
        (value === null || value === void 0 ? void 0 : value.id) && (react_1.default.createElement(DocumentDrawer, { onSave: updateUpload })),
        react_1.default.createElement(ListDrawer, { onSelect: swapUpload })));
};
exports.default = Element;
