// payload/dist/admin/components/elements/ListDrawer/DrawerContent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListDrawerContent = void 0;
const react_1 = __importStar(require("react"));
const modal_1 = require("@faceless-ui/modal");
const react_i18next_1 = require("react-i18next");
const getTranslation_1 = require("../../../../utilities/getTranslation");
const Config_1 = require("../../utilities/Config");
const Auth_1 = require("../../utilities/Auth");
const DocumentInfo_1 = require("../../utilities/DocumentInfo");
const RenderCustomComponent_1 = __importDefault(require("../../utilities/RenderCustomComponent"));
const usePayloadAPI_1 = __importDefault(require("../../../hooks/usePayloadAPI"));
const Default_1 = __importDefault(require("../../views/collections/List/Default"));
const Label_1 = __importDefault(require("../../forms/Label"));
const ReactSelect_1 = __importDefault(require("../ReactSelect"));
const DocumentDrawer_1 = require("../DocumentDrawer");
const Pill_1 = __importDefault(require("../Pill"));
const X_1 = __importDefault(require("../../icons/X"));
const ViewDescription_1 = __importDefault(require("../ViewDescription"));
const getInitialColumns_1 = __importDefault(require("../../views/collections/List/getInitialColumns"));
const buildColumns_1 = __importDefault(require("../../views/collections/List/buildColumns"));
const formatFields_1 = __importDefault(require("../../views/collections/List/formatFields"));
const Preferences_1 = require("../../utilities/Preferences");
const _1 = require(".");
const buildColumns = ({ collectionConfig, columns, onSelect, t, }) => (0, buildColumns_1.default)({
    collection: collectionConfig,
    columns,
    t,
    cellProps: [{
            link: false,
            onClick: ({ collection, rowData }) => {
                if (typeof onSelect === 'function') {
                    onSelect({
                        docID: rowData.id,
                        collectionConfig: collection,
                    });
                }
            },
            className: `${_1.baseClass}__first-cell`,
        }],
});
const shouldIncludeCollection = ({ coll: { admin: { enableRichTextRelationship }, upload, slug, }, uploads, collectionSlugs, }) => (enableRichTextRelationship && ((uploads && Boolean(upload)) || (collectionSlugs === null || collectionSlugs === void 0 ? void 0 : collectionSlugs.includes(slug))));
const DrawerContent = ({ drawerSlug, onSelect, customHeader, collectionSlugs, uploads, selectedCollection, enabledCollectionConfigs, filterOptions, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    const { t, i18n } = (0, react_i18next_1.useTranslation)(['upload', 'general']);
    const { permissions } = (0, Auth_1.useAuth)();
    const { getPreference, setPreference } = (0, Preferences_1.usePreferences)();
    const { isModalOpen, closeModal } = (0, modal_1.useModal)();
    const [limit, setLimit] = (0, react_1.useState)();
    const [sort, setSort] = (0, react_1.useState)(null);
    const [page, setPage] = (0, react_1.useState)(1);
    const [where, setWhere] = (0, react_1.useState)(null);
    const { serverURL, routes: { api }, collections } = (0, Config_1.useConfig)();
    const [selectedCollectionConfig, setSelectedCollectionConfig] = (0, react_1.useState)(() => {
        let initialSelection;
        if (selectedCollection) {
            // if passed a selection, find it and check if it's enabled
            const foundSelection = collections.find(({ slug }) => slug === selectedCollection);
            if (foundSelection && shouldIncludeCollection({ coll: foundSelection, uploads, collectionSlugs })) {
                initialSelection = foundSelection;
            }
        }
        else {
            // return the first one that is enabled
            initialSelection = collections.find((coll) => shouldIncludeCollection({ coll, uploads, collectionSlugs }));
        }
        return initialSelection;
    });
    const [selectedOption, setSelectedOption] = (0, react_1.useState)(() => (selectedCollectionConfig ? { label: (0, getTranslation_1.getTranslation)(selectedCollectionConfig.labels.singular, i18n), value: selectedCollectionConfig.slug } : undefined));
    const [fields, setFields] = (0, react_1.useState)(() => (0, formatFields_1.default)(selectedCollectionConfig, t));
    const [tableColumns, setTableColumns] = (0, react_1.useState)(() => {
        const initialColumns = (0, getInitialColumns_1.default)(fields, selectedCollectionConfig.admin.useAsTitle, selectedCollectionConfig.admin.defaultColumns);
        return buildColumns({
            collectionConfig: selectedCollectionConfig,
            columns: initialColumns,
            t,
            onSelect,
        });
    });
    // allow external control of selected collection, same as the initial state logic above
    (0, react_1.useEffect)(() => {
        let newSelection;
        if (selectedCollection) {
            // if passed a selection, find it and check if it's enabled
            const foundSelection = collections.find(({ slug }) => slug === selectedCollection);
            if (foundSelection && shouldIncludeCollection({ coll: foundSelection, uploads, collectionSlugs })) {
                newSelection = foundSelection;
            }
        }
        else {
            // return the first one that is enabled
            newSelection = collections.find((coll) => shouldIncludeCollection({ coll, uploads, collectionSlugs }));
        }
        setSelectedCollectionConfig(newSelection);
    }, [selectedCollection, collectionSlugs, uploads, collections, onSelect, t]);
    const activeColumnNames = tableColumns.map(({ accessor }) => accessor);
    const stringifiedActiveColumns = JSON.stringify(activeColumnNames);
    const preferenceKey = `${selectedCollectionConfig.slug}-list`;
    // this is the 'create new' drawer
    const [DocumentDrawer, DocumentDrawerToggler, { drawerSlug: documentDrawerSlug, },] = (0, DocumentDrawer_1.useDocumentDrawer)({
        collectionSlug: selectedCollectionConfig.slug,
    });
    (0, react_1.useEffect)(() => {
        if (selectedOption) {
            setSelectedCollectionConfig(collections.find(({ slug }) => selectedOption.value === slug));
        }
    }, [selectedOption, collections]);
    const collectionPermissions = (_a = permissions === null || permissions === void 0 ? void 0 : permissions.collections) === null || _a === void 0 ? void 0 : _a[selectedCollectionConfig === null || selectedCollectionConfig === void 0 ? void 0 : selectedCollectionConfig.slug];
    const hasCreatePermission = (_b = collectionPermissions === null || collectionPermissions === void 0 ? void 0 : collectionPermissions.create) === null || _b === void 0 ? void 0 : _b.permission;
    // If modal is open, get active page of upload gallery
    const isOpen = isModalOpen(drawerSlug);
    const apiURL = isOpen ? `${serverURL}${api}/${selectedCollectionConfig.slug}` : null;
    const [cacheBust, dispatchCacheBust] = (0, react_1.useReducer)((state) => state + 1, 0); // used to force a re-fetch even when apiURL is unchanged
    const [{ data, isError }, { setParams }] = (0, usePayloadAPI_1.default)(apiURL, {});
    const moreThanOneAvailableCollection = enabledCollectionConfigs.length > 1;
    (0, react_1.useEffect)(() => {
        const params = {};
        if (page)
            params.page = page;
        params.where = {
            ...where ? { ...where } : {},
            ...(filterOptions === null || filterOptions === void 0 ? void 0 : filterOptions[selectedCollectionConfig.slug]) ? {
                ...filterOptions[selectedCollectionConfig.slug],
            } : {},
        };
        if (sort)
            params.sort = sort;
        if (limit)
            params.limit = limit;
        if (cacheBust)
            params.cacheBust = cacheBust;
        setParams(params);
    }, [setParams, page, sort, where, limit, cacheBust, filterOptions, selectedCollectionConfig]);
    (0, react_1.useEffect)(() => {
        const syncColumnsFromPrefs = async () => {
            const currentPreferences = await getPreference(preferenceKey);
            const newFields = (0, formatFields_1.default)(selectedCollectionConfig, t);
            setFields(newFields);
            const initialColumns = (0, getInitialColumns_1.default)(newFields, selectedCollectionConfig.admin.useAsTitle, selectedCollectionConfig.admin.defaultColumns);
            setTableColumns(buildColumns({
                collectionConfig: selectedCollectionConfig,
                columns: (currentPreferences === null || currentPreferences === void 0 ? void 0 : currentPreferences.columns) || initialColumns,
                t,
                onSelect,
            }));
        };
        syncColumnsFromPrefs();
    }, [t, getPreference, preferenceKey, onSelect, selectedCollectionConfig]);
    (0, react_1.useEffect)(() => {
        const newPreferences = {
            limit,
            sort,
            columns: JSON.parse(stringifiedActiveColumns),
        };
        setPreference(preferenceKey, newPreferences);
    }, [sort, limit, stringifiedActiveColumns, setPreference, preferenceKey]);
    const setActiveColumns = (0, react_1.useCallback)((columns) => {
        setTableColumns(buildColumns({
            collectionConfig: selectedCollectionConfig,
            columns,
            t,
            onSelect,
        }));
    }, [selectedCollectionConfig, t, onSelect]);
    const onCreateNew = (0, react_1.useCallback)(({ doc }) => {
        if (typeof onSelect === 'function') {
            onSelect({
                docID: doc.id,
                collectionConfig: selectedCollectionConfig,
            });
        }
        dispatchCacheBust();
        closeModal(documentDrawerSlug);
        closeModal(drawerSlug);
    }, [closeModal, documentDrawerSlug, drawerSlug, onSelect, selectedCollectionConfig]);
    if (!selectedCollectionConfig || isError) {
        return null;
    }
    return (react_1.default.createElement(react_1.Fragment, null,
        react_1.default.createElement(DocumentInfo_1.DocumentInfoProvider, { collection: selectedCollectionConfig },
            react_1.default.createElement(RenderCustomComponent_1.default, { DefaultComponent: Default_1.default, CustomComponent: (_e = (_d = (_c = selectedCollectionConfig === null || selectedCollectionConfig === void 0 ? void 0 : selectedCollectionConfig.admin) === null || _c === void 0 ? void 0 : _c.components) === null || _d === void 0 ? void 0 : _d.views) === null || _e === void 0 ? void 0 : _e.List, componentProps: {
                    collection: {
                        ...selectedCollectionConfig,
                        fields,
                    },
                    customHeader: (react_1.default.createElement("header", { className: `${_1.baseClass}__header` },
                        react_1.default.createElement("div", { className: `${_1.baseClass}__header-wrap` },
                            react_1.default.createElement("div", { className: `${_1.baseClass}__header-content` },
                                react_1.default.createElement("h2", { className: `${_1.baseClass}__header-text` }, !customHeader ? (0, getTranslation_1.getTranslation)((_f = selectedCollectionConfig === null || selectedCollectionConfig === void 0 ? void 0 : selectedCollectionConfig.labels) === null || _f === void 0 ? void 0 : _f.plural, i18n) : customHeader),
                                hasCreatePermission && (react_1.default.createElement(DocumentDrawerToggler, { className: `${_1.baseClass}__create-new-button` },
                                    react_1.default.createElement(Pill_1.default, null, t('general:createNew'))))),
                            react_1.default.createElement("button", { type: "button", onClick: () => {
                                    closeModal(drawerSlug);
                                }, className: `${_1.baseClass}__header-close` },
                                react_1.default.createElement(X_1.default, null))),
                        ((_g = selectedCollectionConfig === null || selectedCollectionConfig === void 0 ? void 0 : selectedCollectionConfig.admin) === null || _g === void 0 ? void 0 : _g.description) && (react_1.default.createElement("div", { className: `${_1.baseClass}__sub-header` },
                            react_1.default.createElement(ViewDescription_1.default, { description: selectedCollectionConfig.admin.description }))),
                        moreThanOneAvailableCollection && (react_1.default.createElement("div", { className: `${_1.baseClass}__select-collection-wrap` },
                            react_1.default.createElement(Label_1.default, { label: t('selectCollectionToBrowse') }),
                            react_1.default.createElement(ReactSelect_1.default, { className: `${_1.baseClass}__select-collection`, value: selectedOption, onChange: setSelectedOption, options: enabledCollectionConfigs.map((coll) => ({ label: (0, getTranslation_1.getTranslation)(coll.labels.singular, i18n), value: coll.slug })) }))))),
                    data,
                    limit: limit || ((_j = (_h = selectedCollectionConfig === null || selectedCollectionConfig === void 0 ? void 0 : selectedCollectionConfig.admin) === null || _h === void 0 ? void 0 : _h.pagination) === null || _j === void 0 ? void 0 : _j.defaultLimit),
                    setLimit,
                    tableColumns,
                    setColumns: setActiveColumns,
                    setSort,
                    newDocumentURL: null,
                    hasCreatePermission,
                    columnNames: activeColumnNames,
                    disableEyebrow: true,
                    modifySearchParams: false,
                    onCardClick: (doc) => {
                        if (typeof onSelect === 'function') {
                            onSelect({
                                docID: doc.id,
                                collectionConfig: selectedCollectionConfig,
                            });
                        }
                        closeModal(drawerSlug);
                    },
                    disableCardLink: true,
                    handleSortChange: setSort,
                    handleWhereChange: setWhere,
                    handlePageChange: setPage,
                    handlePerPageChange: setLimit,
                } })),
        react_1.default.createElement(DocumentDrawer, { onSave: onCreateNew })));
};
const ListDrawerContent = (props) => {
    const { collectionSlugs, uploads, } = props;
    const { collections } = (0, Config_1.useConfig)();
    const [enabledCollectionConfigs] = (0, react_1.useState)(() => collections.filter((coll) => shouldIncludeCollection({ coll, uploads, collectionSlugs })));
    if (enabledCollectionConfigs.length === 0) {
        return null;
    }
    return (react_1.default.createElement(DrawerContent, { ...props, enabledCollectionConfigs: enabledCollectionConfigs }));
};
exports.ListDrawerContent = ListDrawerContent;
